### DISCUSSION TITLE
请教：在一个视频网站开启英文字幕时，是否会导致频繁请求翻译？

### DISCUSSION THREAD

**Discussion by cssmagic** [immersive-translate/immersive-translate#1549] [Help]
### 概述 | Summary

我描述一下我遇到的问题：

1. 我在 [LinkedIn Learning](https://www.linkedin.com/learning/courses) 网站观看视频。开启英文字幕，有时会切换到中文字幕。
2. 打开沉浸式翻译插件，以便翻译视频描述。（选择 OpenAI 服务，输入自备的 API Key。）
3. 观看一段时间。
4. 发现 API 有大量调用消耗，调用时间就是观看视频的时间。

因为看不到具体的 API 调用内容，我暂时只能猜测，是否因为英文字幕不断变化，触发了沉浸式翻译插件实时调用 API 进行翻译（但翻译结果并未显示在页面上）。

当然这只是我的猜测，想请教一下，有没有可能存在这种情况。

### 补充信息 | Additional information

后来用调试工具重现了翻译字幕的请求：

![Screen-Shot-2024-03-20-at-09 47 47pm](https://github.com/immersive-translate/immersive-translate/assets/1231359/bc5ae55d-cbd2-499a-8fb5-756c358724b4)

**Comment by cssmagic** [comment]
我找到一个类似的贴子：[[功能咨询]双语字幕功能是否是同一时刻翻译所有外语字幕](https://github.com/immersive-translate/immersive-translate/discussions/1266)

不过我并没有打开 “使用沉浸式翻译来翻译字幕” 选项（默认就是关闭的）。

![Screen-Shot-2024-03-20-at-09 48 10pm](https://github.com/immersive-translate/immersive-translate/assets/1231359/b0ece4cb-e2f2-4509-9397-1bf69251c6d5)

**Comment by cssmagic** [comment]
我重现了整个过程。在没有打开 “使用沉浸式翻译来翻译字幕” 的情况下，插件还是在不断地发请求翻译字幕。

我提个 bug 吧。

**Comment by keyboard3** [comment]
> 我重现了整个过程。在没有打开 “使用沉浸式翻译来翻译字幕” 的情况下，插件还是在不断地发请求翻译字幕。
> 
> 我提个 bug 吧。

你是不是翻译了网页。网页翻译同时也会触发视频翻译

**Comment by keyboard3** [comment]
> > 我重现了整个过程。在没有打开 “使用沉浸式翻译来翻译字幕” 的情况下，插件还是在不断地发请求翻译字幕。
> > 我提个 bug 吧。
> 
> 你是不是翻译了网页。网页翻译同时也会触发视频翻译

下一版本将视频翻译触发从网页翻译中剥离

**Comment by cssmagic** [comment]
是的，我翻译过网页。

但我重现的时候似乎不需要启动插件翻译网页，就自动会有大量请求。

**Comment by keyboard3** [comment]
> 是的，我翻译过网页。
> 
> 但我重现的时候似乎不需要启动插件翻译网页，就自动会有大量请求。

那不可能，你应该是上一个网页没有关闭，插件在没有翻译触发的情况下不会发送请求

**Comment by cssmagic** [comment]
我打开网页翻译后，然后关闭网页翻译，再刷新网页，还是会不断请求接口去翻译字幕吗？

**Reply by keyboard3** [reply]
关闭网页后，就不会在发送请求

**Comment by cssmagic** [comment]
我刚刚又试了一遍：

1. 浏览器没有打开这个网站的任何页面。
2. 重新打开这个网站，此时没有打开插件的网页翻译。
3. 在调试工具里还是可以看到自动发出翻译字幕的请求。

**Reply by theowenyoung** [reply]
其实我理解这里的逻辑是这样：


只要你开启了 【自动开启双语字幕】，那么在原视频没有提供【对应的中文字幕】的情况下，插件确实会自动翻译该视频的字幕。

@keyboard3 是吧？

那个选项的实际含义是【优先使用沉浸式翻译来翻译字幕】，它的主要场景其实是 Youtube，因为Youtube提供了对应的中文字幕，但是有些用户想使用 Deepl 来翻译这个字幕，那么打开那个选项，就会优先使用 Deepl 去翻译。（这里至少应该修改一下那个 label 的名字】

或者也许我们应该再提供一个选项，叫【如果原视频没有对应的译文字幕，则使用设置的翻译服务自动翻译字幕】，这个选项默认开启。

不知道这样是不是合理一点？

**Reply by keyboard3** [reply]
哦哦 我懂了，就是 @theowenyoung 提到的【自动开启双语字幕】点。其实最大的问题是字幕其实是hook请求方案，必须等待翻译完成才显示。我们的新自托管字幕方案，会根据视频播放进度实时翻译。只不过现在还在部分网站覆盖阶段，还没有所有支持的视频站点覆盖。 
你重点是使用那些视频站点？

**Reply by cssmagic** [reply]
> @keyboard3 
> 你重点是使用那些视频站点？

发生问题的是 [LinkedIn Learning](https://www.linkedin.com/learning/courses) 网站。也不是经常用。

**Reply by cssmagic** [reply]
> @theowenyoung 
> 只要你开启了 【自动开启双语字幕】，那么在原视频没有提供【对应的中文字幕】的情况下，插件确实会自动翻译该视频的字幕。

如果是这样，“使用沉浸式翻译来翻译字幕” 这个开关似乎没有起到实际作用？

另外，我在这个网站观看的视频是有中文字幕的。可能是没识别出来？

**Reply by theowenyoung** [reply]
这个选项主要是针对 youtube的，用户如果需要的话，会优先使用youtube去给他翻译。

对于其他网站，如果没有中文字幕，而用户又开了【自动启用双语字幕】，其实我们在播放视频的时候自动翻译也合理..

**Reply by theowenyoung** [reply]
我还有一点疑问是： 

> 但翻译结果并未显示在页面上

这个不应该啊。 @keyboard3

**Reply by keyboard3** [reply]
![image](https://github.com/immersive-translate/immersive-translate/assets/13177744/18635db5-0385-42f5-9a04-41548226fde9)
显示的啊，翻译生效就会显示。 如果是中文字幕，翻中文是不翻的啊

**Reply by cssmagic** [reply]
我来补充一下：

我看的这是个课程：https://www.linkedin.com/learning/paths/career-essentials-in-generative-ai-by-microsoft-and-linkedin

我确实见到过双语字幕（类似上图），是我从英语字幕切换到中文字幕、再切回英文字幕时出现的，我以为是网站本身的功能——因为双语效果显示几页之后就变成只显示英文了，我以为这是网站特意设计的字幕语种切换的缓冲效果 😂 （现在回想可能是实时翻译跟不上视频播放的速度，所以中文字幕不能同步出现？）

由于双语效果没有长时间持续，我基本上都是切换到中文字幕来看视频的。另外，这套课程都是有中文字幕的（我当时只遇到一集完全没有任何字幕，今天再看好像都有全语种字幕了），理论上应该不需要实时翻译？可能插件没识别出来？

不管怎样，我已经知道预防的办法了。发这个帖也没有责怪的意思。开发一个功能齐全、兼容众多网站的插件非常不容易。我自己也开发过油猴脚本和插件，深知不易。请原谅我的絮叨，后续的优化工作就麻烦两位了 🙏

**Comment by cssmagic** [comment]
> 下一版本将视频翻译触发从网页翻译中剥离

如果以后会解决这个问题，我也就不纠结现在是什么逻辑了。期待早日升级！（一下午不知不觉耗掉 ¥50+ 还是有点怕怕 😂 ）

**Reply by theowenyoung** [reply]
抱歉你的50块...☹️

参考我上面的回复，这个版本解决的问题可能并没有解决你的问题...

@keyboard3

**Reply by keyboard3** [reply]
@theowenyoung 建议送个会员压压惊

**Reply by cssmagic** [reply]
@keyboard3 
谢谢，倒不需要补偿，只是希望引起重视，避免其他用户遇到类似损失。

**Reply by cssmagic** [reply]
@theowenyoung 
我现在最保险的选择就是关掉【自动开启双语字幕】，对吧？

**Reply by theowenyoung** [reply]
是的
